### PR TITLE
JAMES-2543 Migrate  SMTP tests to  junit 5

### PR DESCRIPTION
**Pull request by jeantil**
* upgrades MPT SMTP Tests
* upgrades protocol SMTP tests

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]
Nice, thanks a ton!

**Comment by chibenwa** [comment]
> Nice, thanks a ton!

^^ @jeantil historically it had been mostly @Arsnael contributing on this topic.

**Comment by jeantil** [comment]
last count in https://issues.apache.org/jira/browse/JAMES-2543 was 477, with this we get to 
```
> grep -r  -F 'import org.junit.Test'|wc -l  
392
```
almost there !

**Comment by chibenwa** [comment]
> almost there !

I fear the trickiest place remains ^^ (server/mailet/integration-test for instance is going to be challenging!)

**Review by mbaechler — APPROVED** [review]

**Review by rouazana — APPROVED** [review]
Thank you!
Usually we remove the `public` annotation for the test methods because they are not required. Could you handle it?

**Comment by jeantil** [comment]
>Usually we remove the public annotation for the test methods because they are not required. Could you handle it?
 fixed up  :)

**Comment by rouazana** [comment]
> fixed up :)

Thanks that was quick :) What about SMTPStartTlsResponseTest, DNSRBLHandlerTest and more globally all the files modified in the second commit?

**Comment by jeantil** [comment]
> I fear the trickiest place remains ^^ (server/mailet/integration-test for instance is going to be challenging!)

```
> grep -A1 -rh  "@Rule" server/mailet/integration-testing | sort | uniq                                                                                                  
--
    public AmqpRule amqpRule = new AmqpRule(rabbitMqContainer, EXCHANGE_NAME, ROUTING_KEY);
    public ExpectedException expectedException = ExpectedException.none();
    public final RuleChain chain = RuleChain.outerRule(temporaryFolder).around(rabbitMqContainer).around(amqpRule);
    public RuleChain chain = RuleChain.outerRule(rabbit).around(amqpRule).around(folder);
    public SMTPMessageSender messageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
    public SMTPMessageSender smtpMessageSender = new SMTPMessageSender(DEFAULT_DOMAIN);
    public StaticInputChecker resultChecker = new StaticInputChecker();
    public TemporaryFolder folder = new TemporaryFolder();
    public TemporaryFolder temporaryFolder = new TemporaryFolder();
    public TestIMAPClient messageReader = new TestIMAPClient();
    public TestIMAPClient testIMAPClient = new TestIMAPClient();

```
* AmqpRule -> RabbitMqExtension ?
* ExpectedException -> unsure but I would be very surprised if junit 5 doesn't have an equivalent
* Rule chains are replaced by declaration order
* SMTPMessageSender -> SMTPMessageSenderExtension
* TemporaryFolder -> TemporaryFolderExtension

that leaves :

* StaticInputChecker
* TestIMAPClient

followed by a bit of mindless grunt work to manually replace all the @Rule by the corresponding extensions and then a
mass search/replace :
* import org.junit.After ->  import org.junit.jupiter.api.AfterEach ;
* import org.junit.Before -> import org.junit.jupiter.api.BeforeEach;
* import org.junit.Test -> import org.junit.jupiter.api.Rule;
Followed by a mass cleanup imports on the module (thank you intellij)

this could very well be tagged as an easy contribution for external contributors especially if the missing extensions have been built beforehand by more experiences members :)

**Comment by jeantil** [comment]
@rouazana sorry I had initially focused on mpt and forgot that I extended the modifications to the protocol modules :) actualy I fixed up to the wrong commit because of this. there are more files involved in protocol smtp it would exceed pause time and will have to wait for the next pause .

**Comment by Arsnael** [comment]
> ^^ @jeantil historically it had been mostly @Arsnael contributing on this topic.
Yes but I never took the time to completely finish those. So big thanks to you @jeantil :)

**Comment by jeantil** [comment]
there you go 2 fixups to their respective commits to remove public modifier on test methods

File: mpt/impl/smtp/core/src/main/java/org/apache/james/mpt/smtp/SmtpStarttlsCommandTest.java [hunk 1]
```diff
@@ -33,47 +31,42 @@
...
-    
-    private SmtpHostSystem hostSystem;
+    protected SmtpHostSystem hostSystem;
     private SimpleScriptedTestProtocol scriptedTest;
 
-    @Before
-    public void setUp() throws Exception {
-        hostSystem = createSmtpHostSystem();
+    @BeforeEach
+    public void setUp(SmtpHostSystem hostSystem) throws Exception {
```
Comment (rouazana) [inline mpt/impl/smtp/core/src/main/java/org/apache/james/mpt/smtp/SmtpStarttlsCommandTest.java:38]:
```suggestion
    void setUp(SmtpHostSystem hostSystem) throws Exception {
```

File: server/testing/src/main/java/org/apache/james/utils/FakeSmtpExtension.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                .withCommand("node", "cli", "--listen", "80", "--smtp", smtpPort.toString());
+
+        return new FakeSmtpExtension(container);
+    }
+
+    public static FakeSmtpExtension withDefaultPort() {
+        return withSmtpPort(SMTP_PORT);
+    }
+
+    private static  GenericContainer<?> fakeSmtpContainer() {
```
Comment (rouazana) [inline server/testing/src/main/java/org/apache/james/utils/FakeSmtpExtension.java:70]:
```suggestion
    private static GenericContainer<?> fakeSmtpContainer() {
```

**Review by rouazana — APPROVED** [review]
> there you go 2 fixups to their respective commits to remove public modifier on test methods

thanks :)

**Comment by chibenwa** [comment]
```

[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project protocols-lmtp: Compilation failure: Compilation failure: 
[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] /james-project/protocols/lmtp/src/test/java/org/apache/james/protocols/lmtp/AbstractLMTPServerTest.java:[70,5] method does not override or implement a method from a supertype
[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] /james-project/protocols/lmtp/src/test/java/org/apache/james/protocols/lmtp/AbstractLMTPServerTest.java:[75,5] method does not override or implement a method from a supertype
[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] /james-project/protocols/lmtp/src/test/java/org/apache/james/protocols/lmtp/AbstractLMTPServerTest.java:[81,5] method does not override or implement a method from a supertype
[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] /james-project/protocols/lmtp/src/test/java/org/apache/james/protocols/lmtp/AbstractLMTPServerTest.java:[87,5] method does not override or implement a method from a supertype
[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] /james-project/protocols/lmtp/src/test/java/org/apache/james/protocols/lmtp/AbstractLMTPServerTest.java:[92,5] method does not override or implement a method from a supertype
[61b18c2de411a46bd812334b8830cd36477dcd3c] [ERROR] /james-project/protocols/lmtp/src/test/java/org/apache/james/protocols/lmtp/AbstractLMTPServerTest.java:[127,5] method does not override or implement a method from a supertype
``` 

The build fails as LMTP has a dependency on SMTP

**Comment by chibenwa** [comment]
Hello Jean, I merged this nice work ;-)

**Comment by jeantil** [comment]
@chibenwa thanks, I had not seen the last 2 suggestions of @rouazana (extra space and leftover public on a setup method) did you fix them as part of the merge ? (otherwise i'll do a followup MR to cleanup my mess)

**Comment by chibenwa** [comment]
> I had not seen the last 2 suggestions of @rouazana (extra space and leftover public on a setup method) did you fix them as part of the merge ? (otherwise i'll do a followup MR to cleanup my mess)

I've not been careful enough sorry.

Can you fix them in another PR?

**Comment by jeantil** [comment]
>Can you fix them in another PR?

yes

**Pull request closed without merging**